File: ios/OlaKai/Data/LiveStreamResolver.swift
import Foundation

/// Turns a YouTube channel into the video id it is currently broadcasting.
///
/// A catalog cannot pin a video id — operators restart their stream and the old
/// id dies — and `embed/live_stream?channel=` no longer resolves. So the
/// channel's own `/live` page is the source of truth.
actor LiveStreamResolver {

    private struct Entry {
        let videoId: String?
        let at: Date
    }

    private var cache: [String: Entry] = [:]

    /// Streams run for days; re-checking every fifteen minutes is plenty.
    private let ttl: TimeInterval = 15 * 60

    private let browserAgent =
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 "
        + "(KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"

    func currentVideoId(channelId: String) async -> String? {
        guard !channelId.isEmpty else { return nil }

        if let entry = cache[channelId], Date().timeIntervalSince(entry.at) < ttl {
            return entry.videoId
        }

        let resolved = await fetchVideoId(channelId: channelId)
        cache[channelId] = Entry(videoId: resolved, at: Date())
        return resolved
    }

    private func fetchVideoId(channelId: String) async -> String? {
        guard let url = URL(string: "https://www.youtube.com/channel/\(channelId)/live") else {
            return nil
        }
        var request = URLRequest(url: url)
        request.setValue(browserAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("en-US,en;q=0.9", forHTTPHeaderField: "Accept-Language")
        // Without a consent cookie European requests land on the consent wall.
        request.setValue("CONSENT=YES+1; SOCS=CAI", forHTTPHeaderField: "Cookie")

        guard
            let (data, _) = try? await URLSession.shared.data(for: request),
            let html = String(data: data, encoding: .utf8)
        else { return nil }

        return Self.extract(from: html)
    }

    /// Only ids that came from a watch page are trusted.
    ///
    /// When a channel is offline YouTube serves the ordinary channel page, whose
    /// first `"videoId"` is a recommendation. Putting a random video in a surf
    /// cam tile is worse than showing a stale one, so there is deliberately no
    /// loose fallback here.
    static func extract(from html: String) -> String? {
        let patterns = [
            #"<link rel="canonical" href="https://www\.youtube\.com/watch\?v=([\w-]{11})""#,
            #""videoDetails":\{[^}]*?"videoId":"([\w-]{11})""#,
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { continue }
            let range = NSRange(html.startIndex..., in: html)
            if let match = regex.firstMatch(in: html, range: range),
               let idRange = Range(match.range(at: 1), in: html) {
                return String(html[idRange])
            }
        }
        return nil
    }
}

File: app/src/main/java/com/olakai/app/data/youtube/LiveStreamResolver.kt
package com.olakai.app.data.youtube

import android.util.Log
import com.olakai.app.data.Http
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * Turns a YouTube channel into the video id it is currently broadcasting.
 *
 * A cam catalog cannot pin a video id: operators restart their stream and the
 * old id dies. It cannot rely on the `embed/live_stream?channel=` endpoint
 * either -- that resolves inconsistently. So the channel's own `/live` page,
 * which YouTube resolves to the current broadcast, is the source of truth, and
 * the app embeds the concrete video id it names.
 */
class LiveStreamResolver {

    private data class Entry(val videoId: String?, val atMillis: Long)

    private val cache = mutableMapOf<String, Entry>()
    private val lock = Mutex()

    suspend fun currentVideoId(channelId: String): String? {
        if (channelId.isBlank()) return null

        lock.withLock {
            cache[channelId]?.let { entry ->
                if (System.currentTimeMillis() - entry.atMillis < TTL_MS) return entry.videoId
            }
        }

        val resolved = withContext(Dispatchers.IO) {
            // YouTube serves different markup to non-browser agents, and the
            // extraction depends on the markup, so ask as a browser would.
            runCatching {
                extract(
                    Http.getString(
                        url = liveUrl(channelId),
                        userAgent = BROWSER_UA,
                        // Without a consent cookie European requests get
                        // redirected to the consent wall instead of the page.
                        headers = mapOf(
                            "Accept-Language" to "en-US,en;q=0.9",
                            "Cookie" to "CONSENT=YES+1; SOCS=CAI",
                        ),
                    ),
                )
            }
                .onFailure { Log.w(TAG, "resolve failed for $channelId: ${it.message}") }
                .getOrNull()
        }

        lock.withLock { cache[channelId] = Entry(resolved, System.currentTimeMillis()) }
        return resolved
    }

    private fun liveUrl(channelId: String) =
        "https://www.youtube.com/channel/$channelId/live"

    /**
     * Only trust ids that came from a watch page.
     *
     * `videoDetails` and the canonical link exist only when the channel actually
     * resolved to a broadcast. When it is offline YouTube serves the ordinary
     * channel page, whose first `"videoId"` is a recommendation -- taking that
     * would put a random video in a surf cam tile, which is worse than showing
     * the catalogued one. So there is no loose fallback here on purpose.
     */
    private fun extract(html: String): String? =
        CANONICAL.find(html)?.groupValues?.get(1)
            ?: PLAYER_VIDEO_ID.find(html)?.groupValues?.get(1)

    private companion object {
        const val TAG = "LiveStreamResolver"

        /** Streams run for days; re-checking every 15 minutes is plenty. */
        const val TTL_MS = 15 * 60 * 1000L

        const val BROWSER_UA =
            "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 (KHTML, like Gecko) " +
                "Chrome/125.0.0.0 Mobile Safari/537.36"

        val CANONICAL = Regex("""<link rel="canonical" href="https://www\.youtube\.com/watch\?v=([\w-]{11})"""")
        val PLAYER_VIDEO_ID = Regex(""""videoDetails":\{[^}]*?"videoId":"([\w-]{11})"""")
    }
}
